File: java/strategy/DuckSimulator/client/Duck.java
package client;

import behavior.fly.FlyBehavior;
import behavior.quack.QuackBehavior;

public class Duck {

    private final String name;

    FlyBehavior flyBehavior;
    QuackBehavior quackBehavior;

    public Duck(String name) { this.name = name; }

    public void display() { System.out.println("I am a " + name); }

    public void swim() { System.out.println(name + " is swimming"); }

    public void performFly() { flyBehavior.fly(); }

    public void performQuack() { quackBehavior.quack(); }

    public void setFlyBehavior(FlyBehavior flyBehavior) {
        this.flyBehavior = flyBehavior;
    }

    public void setQuackBehavior(QuackBehavior quackBehavior) {
        this.quackBehavior = quackBehavior;
    }
}

File: kotlin/strategy/DuckSimulator/client/Duck.kt
package client

import behavior.fly.FlyBehavior
import behavior.quack.QuackBehavior

open class Duck(private var duckName: String) {

    protected open var flyBehavior: FlyBehavior? = null
    protected open var quackBehavior: QuackBehavior? = null

    fun display() { println("I am a $duckName") }

    fun swim() { println("$duckName is swimming") }

    fun performFly() { flyBehavior!!.fly() }

    fun performQuack() { quackBehavior!!.quack() }

}